Classes/Private/Manager/Audio/OCTAudioQueue.m

```diff
@@ -0,0 +1,418 @@
+    CFStringRef unique = NULL;
+    size = sizeof(unique);
+    ok = _AudioObjectGetPropertyData(devID, &address, 0, NULL, &size, &unique);
+    if (ok != kAudioHardwareNoError) {
+        DDLogCError(@"failed AudioObjectGetPropertyData for selected device: %d! Crash may or may not be imminent", ok);
+        if (err) {
+            *err = OCTErrorFromCoreAudioCode(ok);
+        }
+        return nil;
+    }
+
+    return (__bridge NSString *)unique;
+}
+#endif
+
+@interface OCTAudioQueue ()
+
+// use this to track what nil means in terms of audio device
+@property (assign, nonatomic) BOOL isOutput;
+@property (assign, nonatomic) AudioStreamBasicDescription streamFmt;
+@property (assign, nonatomic) AudioQueueRef audioQueue;
+@property (assign, nonatomic) TPCircularBuffer buffer;
+@property (assign, nonatomic) BOOL running;
+
+@end
+
+@implementation OCTAudioQueue {
+    AudioQueueBufferRef _AQBuffers[kNumberOfAudioQueueBuffers];
+}
+
+- (instancetype)initWithInputDeviceID:(NSString *)devID error:(NSError **)error
+{
+#if TARGET_OS_IPHONE
+    AVAudioSession *session = [AVAudioSession sharedInstance];
+    _streamFmt.mSampleRate = session.sampleRate;
+#else
+    _streamFmt.mSampleRate = kDefaultSampleRate;
+#endif
+    _streamFmt.mFormatID = kAudioFormatLinearPCM;
+    _streamFmt.mFormatFlags = kLinearPCMFormatFlagIsSignedInteger;
+    _streamFmt.mChannelsPerFrame = kNumberOfInputChannels;
+    _streamFmt.mBytesPerFrame = kBytesPerSample * kNumberOfInputChannels;
+    _streamFmt.mBitsPerChannel = kBitsPerByte * kBytesPerSample;
+    _streamFmt.mFramesPerPacket = kFramesPerPacket;
+    _streamFmt.mBytesPerPacket = kBytesPerSample * kNumberOfInputChannels * kFramesPerPacket;
+    _isOutput = NO;
+    _deviceID = devID;
+
+    TPCircularBufferInit(&_buffer, kBufferLength);
+    OSStatus res = 0;
+    if ((res = [self createAudioQueue]) != 0) {
+        if (error) {
+            *error = OCTErrorFromCoreAudioCode(res);
+        }
+        return nil;
+    }
+
+    return self;
+}
+
+- (instancetype)initWithOutputDeviceID:(NSString *)devID error:(NSError **)error
+{
+    _streamFmt.mSampleRate = kDefaultSampleRate;
+    _streamFmt.mFormatID = kAudioFormatLinearPCM;
+    _streamFmt.mFormatFlags = kLinearPCMFormatFlagIsSignedInteger | kLinearPCMFormatFlagIsPacked;
+    _streamFmt.mChannelsPerFrame = kNumberOfInputChannels;
+    _streamFmt.mBytesPerFrame = kBytesPerSample * kNumberOfInputChannels;
+    _streamFmt.mBitsPerChannel = kBitsPerByte * kBytesPerSample;
+    _streamFmt.mFramesPerPacket = kFramesPerPacket;
+    _streamFmt.mBytesPerPacket = kBytesPerSample * kNumberOfInputChannels * kFramesPerPacket;
+    _isOutput = YES;
+    _deviceID = devID;
+
+    TPCircularBufferInit(&_buffer, kBufferLength);
+    OSStatus res = 0;
+    if ((res = [self createAudioQueue]) != 0) {
+        if (error) {
+            *error = OCTErrorFromCoreAudioCode(res);
+        }
+        return nil;
+    }
+
+    return self;
```

Comment:
You can avoid code duplication by moving shared code to private init method. Then you can call it from both inits. Something like that

```objc
- (instancetype)initWithDeviceID:(NSString *)devID
                        isOutput:(BOOL)isOutput
                      sampleRate:(Float64)sampleRate
                     formatFlags:(AudioFormatFlags)formatFlags
                           error:(NSError **)error
```